applications/vanilla/models/class.discussionmodel.php

```diff
@@ -1110,39 +1110,31 @@ public function calculate(&$discussion) {
 
         if (!property_exists($discussion, 'Read')) {
             $discussion->Read = !(bool)$discussion->CountUnreadComments;
-            if (!is_null($category['DateMarkedRead'])) {
+            if ($category && !is_null($category['DateMarkedRead'])) {
                 // If the category was marked explicitly read at some point, see if that applies here
-                if ($category['DateMarkedRead'] > $discussion->DateLastComment) {
-                    $discussion->Read = true;
-                }
-
-                if ($discussion->Read) {
-                    $discussion->CountUnreadComments = 0;
-                }
-            }
-
-            // Discussions are always unread to guests.
-            if (!Gdn::session()->isValid()) {
-                $discussion->Read = false;
+                $discussion->DateLastViewed = self::determineLastViewedDate($discussion->DateLastComment, $category['DateMarkedRead']);
```

Comment:
This is wrong. If I haven't session'd then I don't have a `$category['DateMarkedRead']`.